mlflow/tracing/types/wrapper.py

```diff
@@ -1,32 +1,89 @@
 import json
 import logging
-from functools import lru_cache
-from typing import Any, Dict, Optional, Union
-
-from opentelemetry import trace as trace_api
-
-from mlflow.entities import Span, SpanContext, SpanEvent, SpanStatus, SpanType, TraceStatus
+from dataclasses import asdict
+from functools import lru_cache, wraps
+from typing import Any, Dict, List, Optional, Union
+
+from opentelemetry.sdk.trace import Event as OTelEvent
+from opentelemetry.sdk.trace import ReadableSpan as OTelReadableSpan
+from opentelemetry.trace import Span as OTelSpan
+
+from mlflow.entities.span import SpanType
+from mlflow.entities.span_event import SpanEvent
+from mlflow.entities.span_status import SpanStatus
+from mlflow.entities.trace_status import TraceStatus
+from mlflow.exceptions import MlflowException
+from mlflow.protos.databricks_pb2 import INVALID_PARAMETER_VALUE
 from mlflow.tracing.types.constant import SpanAttributeKey
-from mlflow.tracing.utils import TraceJSONEncoder, format_span_id, format_trace_id
+from mlflow.tracing.utils import (
+    TraceJSONEncoder,
+    build_otel_context,
+    decode_span_id,
+    decode_trace_id,
+    format_span_id,
+    format_trace_id,
+)
 
 _logger = logging.getLogger(__name__)
 
 
-class MlflowSpanWrapper:
+def live_span_only(func):
```

Comment:
@B-Step62 With this approach, users need to reason about when a span is "live" vs "not live" and risk unexpected exceptions if they accidentally call a setter on a non-live trace. Instead of defining live and not live methods on a single object and making users reason about when they can be called, can we use a bit of inheritance here to make this simpler for users?

```
class Span:
    # Defines all internal span properties
    # Defines all "not live" user-facing APIs
```

```
class ActiveSpan(Span):
    # Defines live methods on top of the Span
```

This is similar to MLflow's `ActiveRun` concept.

Reply:
@dbczumar Yeah using inheritance makes sense on second thought. and it aligns with what OTel does. We should probably use `LiveSpan` or `MutableSpan` because "active" is different concept https://github.com/mlflow/mlflow/pull/11736#discussion_r1568271337.

Reply:
`LiveSpan` makes sense :)